calyx/opt/src/passes_experimental/metadata_table_gen.rs

```diff
@@ -55,57 +75,168 @@ impl Named for Metadata {
+            let pos = self
+                .src_table
+                .push_position(*fid, LineNum::new(span.0 as u32));
+            // add tag to group attributes
+            attr.insert_set(calyx_frontend::SetAttr::Pos, pos.value());
+            //dbg!(&self.src_table);
+        });
+        //dbg!(&self.src_table);
+
+        Ok(Action::Continue)
     }
-}
 
-#[cfg(test)]
-mod tests {
-    use std::path::PathBuf;
-
-    use calyx_ir::Id;
-
-    use crate::passes_experimental::metadata_table_gen::Metadata;
-    #[test]
-    fn test_metadata() {
-        let mut data = Metadata::new();
-        let empt_string = data.to_string();
-        assert_eq!(empt_string, "");
-
-        let path = PathBuf::from("/temp/path/for/testing.futil");
-        data.add_entry(
-            Id::from("main"),
-            Id::from("group_1"),
-            (12, 16),
-            path.clone(),
-        );
-        data.add_entry(Id::from("main"), Id::from("group_2"), (23, 28), path);
-        let test_string = data.to_string();
-        assert_eq!(
-            test_string,
-            "main.group_1: /temp/path/for/testing.futil 12-16\nmain.group_2: /temp/path/for/testing.futil 23-28\n"
-        )
+    fn finish_context(&mut self, ctx: &mut calyx_ir::Context) -> VisResult {
+        //dbg!(&self.src_table);
+        ctx.source_info_table = Some(std::mem::take(&mut self.src_table));
+        //dbg!(&ctx.source_info_table);
+        Ok(Action::Continue)
+    }
+
+    // control nodes
+    fn start_seq(
+        &mut self,
+        s: &mut calyx_ir::Seq,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &calyx_ir::LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        self.add_control_node(s);
+        Ok(Action::Continue)
+    }
+
+    fn start_par(
+        &mut self,
+        s: &mut calyx_ir::Par,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &calyx_ir::LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        self.add_control_node(s);
+        Ok(Action::Continue)
+    }
+
+    fn start_if(
+        &mut self,
+        s: &mut calyx_ir::If,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &calyx_ir::LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        self.add_control_node(s);
+        Ok(Action::Continue)
+    }
+
+    fn start_while(
+        &mut self,
+        s: &mut calyx_ir::While,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &calyx_ir::LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        self.add_control_node(s);
+        Ok(Action::Continue)
+    }
+
+    fn start_repeat(
+        &mut self,
+        s: &mut calyx_ir::Repeat,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &calyx_ir::LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        self.add_control_node(s);
+        Ok(Action::Continue)
     }
```

Comment:
We seem to be missing enable! We also want to implement `enable` so that group
enables each receive their own position tag.